### PR TITLE
refactor writeback: emit normalization errors with new solver

### PR DESCRIPTION
**Pull request by lcnr**
implements https://github.com/rust-lang/rust/pull/118725#discussion_r1419824030

r? @compiler-errors @BoxyUwU whoever comes first n stuff

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/writeback.rs [hunk 1]
```diff
@@ -768,64 +768,73 @@ impl<'cx, 'tcx> Resolver<'cx, 'tcx> {
...
-}
 
-///////////////////////////////////////////////////////////////////////////
-// During type check, we store promises with the result of trait
-// lookup rather than the actual results (because the results are not
-// necessarily available immediately). These routines unwind the
-// promises. It is expected that we will have already reported any
-// errors that may be encountered, so if the promises store an error,
-// a dummy result is returned.
+    fn fold_predicate(&mut self, predicate: ty::Predicate<'tcx>) -> ty::Predicate<'tcx> {
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/writeback.rs:831]: When do we resolve predicates during writeback?

File: compiler/rustc_hir_typeck/src/writeback.rs [hunk 2]
```diff
@@ -768,64 +768,73 @@ impl<'cx, 'tcx> Resolver<'cx, 'tcx> {
...
+    fn handle_term<T>(
+        &mut self,
+        value: T,
+        outer_exclusive_binder: impl FnOnce(T) -> ty::DebruijnIndex,
+        new_err: impl Fn(TyCtxt<'tcx>, ErrorGuaranteed) -> T,
+    ) -> T
+    where
+        T: Into<ty::GenericArg<'tcx>> + Copy,
+        T: TypeFoldable<TyCtxt<'tcx>>,
+        T: TypeSuperFoldable<TyCtxt<'tcx>>,
```
[the remark this answers is not in the document]
Comment (lcnr) [reply compiler/rustc_hir_typeck/src/writeback.rs:781]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TypeckResults.html#structfield.coroutine_interior_predicates

File: compiler/rustc_hir_typeck/src/writeback.rs (hunk 1 above)
Comment (lcnr) [reply compiler/rustc_hir_typeck/src/writeback.rs:831]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TypeckResults.html#structfield.coroutine_interior_predicates

File: compiler/rustc_hir_typeck/src/lib.rs [hunk 3]
```diff
@@ -283,8 +283,6 @@ fn typeck_with_fallback<'tcx>(
 
     fcx.check_asms();
 
-    fcx.infcx.skip_region_resolution();
```
[the remark this answers is not in the document]
Comment (lcnr) [reply compiler/rustc_hir_typeck/src/lib.rs:286]: I think this is fine.

We only did this to enable the use of `fully_resolve` in writeback to also resolve regions. But given that we resolve all regions to `'erased` anyways, this is pretty pointless and `resolve_vars_if_possible` is good enough.

File: compiler/rustc_hir_typeck/src/writeback.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/writeback.rs:829]: Oh right

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**